driver-core/src/main/java/com/datastax/driver/core/NettyCustomizer.java

```diff
@@ -0,0 +1,165 @@
+/*
+ *      Copyright (C) 2012-2014 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import java.lang.reflect.Constructor;
+import java.util.concurrent.ThreadFactory;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.epoll.EpollSocketChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A set of hooks that allow clients to customize the driver's underlying Netty layer.
+ * <p>
+ * Clients that need to plug-in their own logic into the driver's underlying Netty layer can
+ * subclass this class and provide the necessary customization through any of the
+ * publicly available methods.
+ * <p>
+ * Typically, clients would register this class with Cluster.Builder:
+ *
+ * <pre>
+ *     ChannelCustomizer customizer = ...
+ *     Cluster cluster = Cluster.builder()
+ *          .addContactPoint(...)
+ *          .withNettyCustomizer(customizer)
+ *          .build();
+ * </pre>
+ *
+ * <p>
+ * <strong>Warning:</strong> since 2.0.9 and 2.1.4 (see JAVA-538),
+ * the driver is available in two different flavors: with a standard Maven dependency on Netty,
+ * or with a "shaded" (internalized) Netty dependency.
+ * Since this API exposes Netty classes ({@link SocketChannel}, etc.) to client applications,
+ * <em>it should only be extended by clients using the non-shaded
+ * version of driver</em>. <strong>Extending this API with shaded Netty classes is not supported,
```

Comment:
Adding an extra warning about shaded netty + api extension not being supported and being subject to runtime errors in OSGi.